### ISSUE TITLE
Extension cloud function to use transaction

### ISSUE THREAD

**Issue by maksim-sidorenko** [algolia/firestore-algolia-search#200]
Right now the cloud function for extension, even with the `forceDataSync` flag enabled, does not use transactions, so it is still technically possible to have a conflict on multiple concurrent updates to the listened document.

[[function code]](https://github.com/algolia/firestore-algolia-search/blob/8976e65d4be4878b3ffcb9709a9725b38696da87/functions/src/index.ts#L75C1-L119C3)
```ts
const handleUpdateDocument = async (
  before: DocumentSnapshot,
  after: DocumentSnapshot,
  timestamp: Number
) => {
  try {
    const forceDataSync = config.forceDataSync;
    if (forceDataSync) {
      const updatedSnapshot = await after.ref.get();
      const data = await extract(updatedSnapshot, 0);
      logs.updateIndex(updatedSnapshot.id, data);
      logs.info('force sync data: execute saveObject');
      await index.saveObject(data);
    } else {
      if (areFieldsUpdated(config, before, after)) {
        logs.debug('Detected a change, execute indexing');

        const beforeData: DocumentData = await before.data();
        // loop through the after data snapshot to see if any properties were removed
        const undefinedAttrs = Object.keys(beforeData).filter(key => after.get(key) === undefined || after.get(key) === null);
        logs.debug('undefinedAttrs', undefinedAttrs);
        // if no attributes were removed, then use partial update of the record.
        if (undefinedAttrs.length === 0) {
          const data = await extract(after, timestamp);
          logs.updateIndex(after.id, data);
          logs.debug('execute partialUpdateObject');
          await index.partialUpdateObject(data, { createIfNotExists: true });
        }
        // if an attribute was removed, then use save object of the record.
        else {
          const data = await extract(after, 0);

          // delete null value attributes before saving.
          undefinedAttrs.forEach(attr => delete data[attr]);

          logs.updateIndex(after.id, data);
          logs.debug('execute saveObject');
          await index.saveObject(data);
        }
      }
    }
  } catch (e) {
    logs.error(e);
  }
};
```

Can a transactional approach be added to resolve multiple concurrent updates conflicts?

**Comment by smomin** [comment]
Is this a rampant issue or are we solving for a corner case issue?

**Comment by maksim-sidorenko** [comment]
> Is this a rampant issue or are we solving for a corner case issue?

I would say it's more of a corner case issue, not really urgent, but nice to have
